src/ECS/Systems/PathfindingSystem.cpp

```diff
@@ -28,43 +39,557 @@ PathfindingSystem& PathfindingSystem::instance()
+					// http://paulbourke.net/geometry/circlesphere/
+					auto& fixed = registry.Get<const Fixed>(*iter);
+					const auto d2 = glm::distance2(fixed.boundingCenter, obstacleFixed.boundingCenter);
+					const auto d = glm::sqrt(d2);
+
+					// Vanilla bug: Scaling is already applied to boundingRadius, but they apply scale again
+					const float fixedScale = glm::compMax(registry.Get<const Transform>(*iter).scale);
+					const float obstacleScale = glm::compMax(registry.Get<const Transform>(reference.entity).scale);
+					const auto r0 = fixed.boundingRadius * fixedScale;
+					const auto r1 = obstacleFixed.boundingRadius * obstacleScale;
+
+					const auto r02 = r0 * r0;
+					const auto r12 = r1 * r1;
+					const auto p0 = fixed.boundingCenter;
+					const auto p1 = obstacleFixed.boundingCenter;
+					const auto a = (r02 - r12 + d2) / (2.0f * d); // first circle to intersection midpoint
+					const auto h = glm::sqrt(r02 - a * a);        // half height of intersection area
+					const auto p2 = p0 + a * (p1 - p0) / d;       // midpoint of overlap
+					const auto diff = p1 - p0;
+					const auto difft = glm::vec2(diff.y, -diff.x); // 90 degree rotation
+					const auto p3 = p2 + h * difft / d;
+					const auto p4 = p2 - h * difft / d;
+
+					const auto v0 = p3 - obstacleFixed.boundingCenter;
+					const auto v1 = p4 - obstacleFixed.boundingCenter;
+					const auto n0 = glm::normalize(v0);
+					const auto n1 = glm::normalize(v1);
+					const auto dp0 = glm::dot(n0, circleNormal);
+					const auto dp1 = glm::dot(n1, circleNormal);
+					const auto cp0 = glm::cross(glm::vec3(n0, 0.0f), glm::vec3(circleNormal, 0.0f)).z;
+					const auto cp1 = glm::cross(glm::vec3(n1, 0.0f), glm::vec3(circleNormal, 0.0f)).z;
+					auto angle0 = glm::acos(dp0);
+					auto angle1 = glm::acos(dp1);
+
+					if ((cp0 > 0.0f) ^ clockwise)
+					{
+						angle0 = 2.0f * glm::pi<float>() - angle0;
+					}
+					if ((cp1 > 0.0f) ^ clockwise)
+					{
+						angle1 = 2.0f * glm::pi<float>() - angle1;
+					}
+
+					const auto t0 = angle0 * 2.0f / 3.0f * r1 / wallHug.speed;
+					const auto t1 = angle1 * 2.0f / 3.0f * r1 / wallHug.speed;
+					int t = static_cast<int>(glm::round(glm::min(t0, t1)));
+
+					assert(t >= 0);
+					if (t < 1)
+					{
+						// We're too close to second circle. Act like we're on the second circle and continue looking forward by
+						// recursively calling function with new obstacle.
+						reference.entity = *iter;
+						OrbitScanForObstacle(entity, clockwise, transform, wallHug);
+					}
+					else if (t < 4)
+					{
+						t = 0;
+					}
+					else
+					{
+						t = glm::min(t, 255);
+					}
+
+					reference.stepsAway = static_cast<uint8_t>(t);
+				}
+				else
+				{
+					// TODO:
+					// if intersect[0].obj is None:  # True
+					//     self.init_steps_xz()
+					//     # self.field_0x78 = 0x10
+					//     self.circle_hug_info.reset(self)
+					//     self.move_state = MoveState.STEP_THROUGH
+					// assert(false);
```

Comment:
Why is this in TODO?

Reply:
I'm not sure if this is needed. It's a corner case.
If we notice some villagers getting stuck. This might be where they get stuck...